### PR TITLE
Weight Standardization method

### PR DESCRIPTION
**Pull request by Landanjs**
Composer algorithm implementation for [weight standardization](https://arxiv.org/abs/1903.10520). 

Positive results when training ResNet-50 on ImageNet and training DeepLabv3+ on ADE20k. Weight standardization on DeepLabv3+ only worked when ignoring the last 15 convolutional layers (i.e. the DeepLabv3+ head). Still not sure why standardizing these layers is harmful...

I was unable to get alternative normalization layers like LayerNorm and GroupNorm to work with weight standardization. Throughput was cut in half and convergence wasn't as good for ResNet-50 on ImageNet. It seemed like training was unstable since there were loss spikes, but ensure why this is happening in our benchmarks, but not in the other scenarios weight standardization is used (i.e. BiT).

[WandB runs](https://wandb.ai/mosaic-ml/weight-standardization), then group by "Name". There are a lot of runs, but the important ones are:
- "dlv3p-ade20k-dice-baseline-nopretrain" (DeepLabv3+ baseline)
- "dlv3p-ade20k-dice-ws_ignored15-nopretrain" (DeepLabv3+ weight standardization and ignore last 15 layers)
- "r50-i1k-baseline" (ResNet50 baseline)
- "r50-i1k-weight-standard" (ResNet50 weight standardization)

### CODE REVIEW

**Comment by eracah** [comment]
nice use of the torch `parametrize` tool!

**Comment by Landanjs** [comment]
@dblalock @dskhudia it added y'all by default, feel free to ignore this or give a fake review. I'll get Austin or Cory to review this

File: composer/algorithms/weight_standardization/README.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+## Suggested Hyperparameters
+
+We found the best performance resulted from setting `n_last_layers_ignore` equal to the number of classification layers that are convolutional. For example, we set `n_last_layers_ignore=0` for training ResNet-50 on ImageNet since the classification is a single linear layer and we set `n_last_layers_ignore=15` for training DeepLabv3+ on ADE20k since the DeepLabv3+ classification head consists of 15 convolution layers.
+
+## Technical Details
+
+For ResNet-50 trained on ImageNet, we measured a +0.08% improvement when using weight standardization. For DeepLabv3+ with a ResNet-101 backbone trained on ADE20k, we measured a +0.7 mIoU improvement when using `n_last_layers_ignore=15`. In addition to the improvements at the end of training, we observed larger improvements early in training and sometimes a decrease in training loss. This suggests the potential for weight standardized models to be trained with more regularization. We have only tested the performance improvement from this method on ResNet architectures.
```
Comment (coryMosaicML) [inline composer/algorithms/weight_standardization/README.md:102]: Do we see this improvement when using pretrained weights, or only when training from scratch?

File: composer/algorithms/weight_standardization/weight_standardization.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+log = logging.getLogger(__name__)
+
+__all__ = ['apply_weight_standardization', 'WeightStandardization']
+
+
+def _standardize_weights(W: torch.Tensor):
+    """Function to standardize the input weight ``W``"""
+    reduce_dims = list(range(1, W.dim()))
+    W_var, W_mean = torch.var_mean(W, dim=reduce_dims, keepdim=True, unbiased=False)
+    return (W - W_mean) / (torch.sqrt(W_var + 1e-10))
```
Comment (coryMosaicML) [inline composer/algorithms/weight_standardization/weight_standardization.py:25]: What do you think about adding an option here to account for the fan in? Not quite full scaled weight standardization, but a step closer. Default behavior would still be vanilla weight standardization, but scaling by `1/sqrt(fan_in)` here might be helpful, but not different enough to warrant being it's own algorithm.

File: composer/algorithms/weight_standardization/weight_standardization.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    Args:
+        module (torch.nn.Module): the torch module whose convolutional weights will be parametrized.
+        n_last_layers_ignore (int, optional): the number of layers at the end of the module to not apply weight standardization.
+            Default: ``0``.
+    """
+    modules_to_parametrize = (nn.Conv1d, nn.Conv2d, nn.Conv3d)
+    # Attempt to symbolically trace a module, so the results of .modules() will be in the order of execution
+    try:
+        module_trace = symbolic_trace(module)
```
Comment (coryMosaicML) [inline composer/algorithms/weight_standardization/weight_standardization.py:46]: Do you know why this might fail? If so, maybe include a bit about it in the warning.

**Review by coryMosaicML — APPROVED** [review]
Overall looks good! One minor question and suggested an optional improvement.

File: composer/algorithms/weight_standardization/README.md (hunk 1 above)
Comment (Landanjs) [reply composer/algorithms/weight_standardization/README.md:106]: Only training from scratch since I haven't tried using pre-trained weights trained with weight standardization. I will update the doc!

File: composer/algorithms/weight_standardization/weight_standardization.py (hunk 3 above)
Comment (Landanjs) [reply composer/algorithms/weight_standardization/weight_standardization.py:46]: This was actually failing with our BERT model, but I did not dig into it 😅

Comment (Landanjs) [reply composer/algorithms/weight_standardization/weight_standardization.py:46]: It looks like this was failing due to unsupported operations in the forward function. Specifically, there is a for-loop with an if-statement in the `forward()` of our huggingface integration which is not support by `symbolic_trace()`. I added more details to the warning. Let me know if it is sufficient!

**Comment by Landanjs** [comment]
@dskhudia or @dblalock looks like I can't merge without a review from one of y'all. Would it be okay to get a stamp approval? Thanks!

**Review by dskhudia — APPROVED** [review]
rubber stamp

File: composer/algorithms/weight_standardization/weight_standardization.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    except:
+        if n_last_layers_ignore > 0:
+            log.warning(
+                textwrap.dedent(f"""\
+                Module could not be symbolically traced likely due to logic in forward() which is not traceble. Modules
```
Comment (dblalock) [inline composer/algorithms/weight_standardization/weight_standardization.py:51]:
```suggestion
                Module could not be symbolically traced likely due to logic in forward() which is not traceable. Modules
```

**Review by dblalock — APPROVED** [review]
Basically rubber stamp as requested. Just found a couple typos / slight opportunities for clarification

File: composer/algorithms/weight_standardization/README.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+## Suggested Hyperparameters
+
+We found the best performance resulted from setting `n_last_layers_ignore` equal to the number of classification layers that are convolutional. For example, we set `n_last_layers_ignore=0` for training ResNet-50 on ImageNet since the classification is a single linear layer and we set `n_last_layers_ignore=15` for training DeepLabv3+ on ADE20k since the DeepLabv3+ classification head consists of 15 convolution layers.
+
+## Technical Details
+
+For ResNet-50 trained on ImageNet, we measured a +0.08% improvement when using weight standardization. For DeepLabv3+ with a ResNet-101 backbone trained on ADE20k, we measured a +0.7 mIoU improvement when using `n_last_layers_ignore=15` and training from scratch i.e. without pre-trained weights. In addition to the improvements at the end of training, we observed larger improvements early in training and sometimes a decrease in training loss. This suggests the potential for weight standardized models to be trained with more regularization. We have only tested the performance improvement from this method on ResNet architectures.
```
[the remark this answers is not in the document]
Comment (Landanjs) [reply composer/algorithms/weight_standardization/README.md:106]: Ambitious wording on my end, I'll change it to "similar accuracy"
